### DISCUSSION TITLE
Creating Deep Links in NextJS using Next/Router

### DISCUSSION THREAD

**Discussion by lloydrichards** [pmndrs/jotai#1836] [Q&A]
👋  Hello!

Im in the process of learning Jotai and having some issues when trying to intergrate the atomWithHash() with Next/Router in order to create Deep Links and be able to save and share configurations between users via URL.

I've created a [simple repo](https://github.com/lloydrichards/DeepLinks-in-Nextjs) for testings and have noticed that the state of the page is not loaded correctly on initial load when passing hash values through the Link components.

**To Recreate**

1.  Clone the repo
2. Run `yarn dev`
3. Go to https:localhost:3000
4. Enter a name and see it displayed in the url hash and on the page. 👍 
5. Copy the URL and paste into a new browser and see the same page. 👍 
6. Click on the Nav link to About (localhost:3000/about)
7. Interact with switch and see the same state saved in url 👍 
8. Click on the nav link to About:Sports (localhost:3000/about#topic="sports")
9. See the page not react 👎 
10. Copy the URL and paste into browser to see the correct loading of the topic data 👍 

**Questions**

- Is there a way for the Next/Router to correctly load atomWithHash?
- Does it require additional setup such as the atomWithHashSync() from `src/utils/atoms.ts`
- How to remove the `%22` when passing strings to the hash?

**Comment by t6adev** [comment]
I created the codesandbox project by the repo, [here](https://codesandbox.io/p/github/t6adev/DeepLinks-in-Nextjs/main?file=%2FREADME.md&workspace=%257B%2522activeFileId%2522%253Anull%252C%2522openFiles%2522%253A%255B%255D%252C%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522gitSidebarPanel%2522%253A%2522COMMIT%2522%252C%2522spaces%2522%253A%257B%2522clf9tmj8s000x3b6l6vdvy31m%2522%253A%257B%2522key%2522%253A%2522clf9tmj8s000x3b6l6vdvy31m%2522%252C%2522name%2522%253A%2522Default%2522%252C%2522devtools%2522%253A%255B%257B%2522key%2522%253A%2522clf9tmj8s000y3b6l8ry5kxbw%2522%252C%2522type%2522%253A%2522PROJECT_SETUP%2522%252C%2522isMinimized%2522%253Afalse%257D%252C%257B%2522type%2522%253A%2522PREVIEW%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A3000%252C%2522key%2522%253A%2522clf9tn2m900a53b6l44vv10kb%2522%252C%2522isMinimized%2522%253Afalse%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522key%2522%253A%2522clf9tn0tr00643b6lxb543v2b%2522%252C%2522isMinimized%2522%253Afalse%257D%255D%257D%257D%252C%2522currentSpace%2522%253A%2522clf9tmj8s000x3b6l6vdvy31m%2522%252C%2522spacesOrder%2522%253A%255B%2522clf9tmj8s000x3b6l6vdvy31m%2522%255D%252C%2522hideCodeEditor%2522%253Afalse%257D)
Let's start with it and I add "another page".

**Reply by lloydrichards** [reply]
I added a branch and PR with a nav link that takes the name variable from the home page and used it to construct the link with the hashed data in the url for the topic.

So if you go to the home page, enter a name, click the About: [name] link, see the page doesn't load the data, but then click the toggle switch (which triggers a rerender) then the name will be populated in the render.

This is kind of frustrating as it works correctly still if you copy the url and load the page directly in another tab or browser 🤷

**Reply by dai-shi** [reply]
@lloydrichards Do you know what's the underlying problem and/or how to fix it?

**Reply by lloydrichards** [reply]
From a [discussion](https://github.com/pmndrs/jotai/issues/739) you had with one of my colleges a few years ago, i think it might have something to do with how the atomWithHash is listening to changes with the router, and thus not being triggered.  I've tried creating a atomWithHashSync function to customise the behaviour with Router from 'next/router' but this still isn't producing results.

My understanding of whats going on under the hood with Jotai and Next/Router is limited, but can easily recreate the unexpected behaviour.

**Reply by dai-shi** [reply]
Your `atomWithHashSync` is a little different from this:
https://github.com/pmndrs/jotai/blob/0c195b091a73ca031889e3f39adae87d95c1eaeb/docs/guides/nextjs.mdx?plain=1#L17-L27

Is it intentional? Should we update docs?
I wonder if anyone would be able to dig into it deeper.

**Reply by lloydrichards** [reply]
The changes are what we have working for another project using Next (12.1) and Jotai (1.4) but since I'm having this issue with the latest versions of Next (13.x) and Jotai (2.0), i want to first understand and fix the issue before suggesting any updates.

**Reply by lloydrichards** [reply]
I did a bit more investigating on this and it turns out that the latest Next13 router does not expose events from the Router like used in the atomWithHashSync example.  If you add a console.log to each listenener then the only one that is ever triggered is the windows.events.on("hashchange").

Also figured out that when you log out the hash value that you get null after the initial route change (twice cause of strict mode) and then when you trigger the component to rerender with the toggle then you get one more null and then the value.  This might be from the componentIsUnMounted but not sure whats going on here.

<img width="1444" alt="Screenshot 2023-03-28 at 15 29 23" src="https://user-images.githubusercontent.com/59728961/228381342-8dafa275-6cb7-4e78-821f-72787d8a8a10.png">

**Reply by dai-shi** [reply]
> it turns out that the latest Next13 router does not expose events from the Router like used in the atomWithHashSync example.

that's news to me. there's something we should do then. help wanted.

**Reply by lloydrichards** [reply]
Have added a PR (#1865) to update the docs with a caution about NextJS routing as well as links to the [roadmap](https://beta.nextjs.org/docs/app-directory-roadmap#planned-features) where you can check on the status of the router hash handling and events Intercepting

**Reply by lloydrichards** [reply]
I was checking is this was still an issue with the release of Next 13.4 and it seems like you no longer need to use the atomWithHasSync() to get hash atoms working 🎉  I've updated the repo i've linked above to the latest and things seem to be working as expected.  Might make another PR later this week to update the docs if you'd like

**Reply by dai-shi** [reply]
Yes, please.

**Reply by lloydrichards** [reply]
I spoke too soon 😮‍💨 I wanted to double double check and found that there are now even weirder behaviours when it comes to moving between app and pages directory as well as when moving between routes using atomWithHash and atom.  Have posted an issue with Next [here](https://github.com/vercel/next.js/issues/49601) as i think this is much more of a next router issue then Jotai, but if this reaches any conclusions i'll be sure to update the docs with what works in next 12 and 13

**Reply by lloydrichards** [reply]
i hold my breath but I think the latest minor release of Next13 ([v13.4.5](https://github.com/vercel/next.js/releases/tag/v13.4.5)) may have fixed the issue 🤞 

>App Router: fix relative query/hash handling in next/link and router push/replace: https://github.com/vercel/next.js/pull/49521

This seems to be the change that has helped with hash syncs and so far i've updated the minimal repo and things look promising.  Will try next week to build some url states on a side project and if things go well then i'll update the documentation for you
